### ISSUE TITLE
Atomicity worries in committing updates from zcf to zoe

### ISSUE THREAD

**Issue by erights** [Agoric/agoric-sdk#1391]
From https://github.com/Agoric/agoric-sdk/pull/1357#discussion_r466041482

HAZARD NOTE: Would be a safety violation in E or Midori. Is a safety violation for us if messages are sent over unreliable connections, which they're not within a SwingSet.

This commits each seat staging from zcf to zoe individually. Here we call the local `commit` for each new allocation. That local commit does a

```js
      E(zoeSeat).replaceAllocation(currentAllocation);
```
for each one. If the zcf vat dies after only some of the new allocations are committed[1], it can cause Zoe to violate even offer safety, because some old payouts can overdraw the pooled purses for new allocations.

This is also a good example why the three way split is still needed. Because the pooled purses are pooled zoe-wide, a bad crash by one zcf vat can cause offer safety violations in *unrelated* contracts.

[1] It is not a safety violation for us because
   * Messages are only released from completed turns. All these messages are *sent* from the same turn, so either all or none will be emitted from this vat.
   * Within a swingset, or (perhaps?) over ibc, as long as the receiving is alive, every message ever sent to the receiver will eventually be delivered. By contrast, in captp over an unreliable transport, some of these commit messages may be delivered and the rest may not be due to a broken connection.

So even though this program happens to work on our current infrastructure, I do not want to define our computational model to support such delicate assumptions. For E and Midori, neither bullet was true. For Waterken, both were true. I think we should keep the first bullet property (@dtribble may disagree) but I do not think we should guarantee the second in general. Our model should include inter-vat partition. That makes this program incorrect in theory even if it happens to work today in practice.

**Comment by erights** [comment]
From https://github.com/Agoric/agoric-sdk/pull/1357#discussion_r466046947

OTOH, if these messages are racing with notification of zcf vat death, could zoe react to a zcf vat death after processing some of these messages but before processing others?

**Comment by dtribble** [comment]
Also from #1357 :

Good point on the analysis. And I agree that we should assume bullet 1 and NOT bullet 2 (eventually).

So for alpha, I think that the guarantee we have is good enough, since payout liveness in the event of vat failure was not in scope. We just wanted the design to be on the path to that.

I also thought about the racing with the vat failure notification. Because of atomicity of send, I would really like to signal vat failure only after all messages sent from the vat have been delivered. That is consistent for example with there still being messages queued up sent for previous, already-committed turns/cranks.

**Comment by erights** [comment]
> Because of atomicity of send, I would really like to signal vat failure only after all messages sent from the vat have been delivered.

Good point. I think we must do that. Thus, my concern in my second comment above is a non-issue. Good!

**Comment by katelynsills** [comment]
Why don't we just make the call from zcf to zoe one call over several seats? That would bypass this issue entirely.

**Comment by erights** [comment]
Yes, that would exactly fix the hazard in the best way. 

We should also see what other batches of messages we're sending from zcf to zoe in one turn, to see if there may be other hidden atomicity assumptions. For example, as you remember from reviewing, my `zcfMint.mintGains` and `.burnLosses` methods seemed to have a pair of messages that needed to be atomic. But it turned out that they only needed to be ordered, and an interleaving or failure between them was not harmful.

As I'm reviewing, I'll keep the possibility of others on my radar.
